### PR TITLE
Raise error when passed invalid file paths

### PR DESCRIPTION
**Pull request by stufro**
#388 

See [comment](https://github.com/crystal-ameba/ameba/issues/388#issuecomment-1651562708) on #388 for my thought process. 

- I had to change `spec/ameba/cli/cmd_spec.cr` to use real files to prevent the error being thrown there.
  - This also involved changing [line 8](https://github.com/crystal-ameba/ameba/compare/master...stufro:388-raise-on-invalid-file-path?expand=1#diff-94ee7d95bcd546992514ff5718c32bf9b79b665e04feb4c75234b199cb524e8dR8) of that spec to use a config file which disables some rules defined in `spec/spec_helper.cr` which always adds errors.

### CODE REVIEW

File: src/ameba/glob_utils.cr [hunk 1]
```diff
@@ -21,6 +21,8 @@ module Ameba
     # ```
     def expand(globs)
       globs.flat_map do |glob|
+        raise ArgumentError.new("No files found matching #{glob}") if Dir[glob].empty?
+
         glob += "/**/*.cr" if File.directory?(glob)
         Dir[glob]
```
Comment (veelenga) [inline src/ameba/glob_utils.cr:27]: the assertion probably needs to be moved one line below to raise on empty folders:

```suggestion
        glob += "/**/*.cr" if File.directory?(glob)
        
        raise ArgumentError.new("No files found matching #{glob}") if Dir[glob].empty?
        
        Dir[glob]
```

Comment (veelenga) [reply src/ameba/glob_utils.cr:27]: Also, could you please add a test for this case?

File: spec/ameba/cli/cmd_spec.cr [hunk 2]
```diff
@@ -43,12 +43,12 @@ module Ameba::Cli
       end
 
       it "defaults rules? flag to false" do
-        c = Cli.parse_args %w(file.cr)
+        c = Cli.parse_args %w(spec/fixtures/passing_ameba.cr)
```
Comment (veelenga) [inline spec/ameba/cli/cmd_spec.cr:46]: let's rename the fixture to `passing_ameba.cr` -> `source.cr`

File: spec/ameba/glob_utils_spec.cr [hunk 3]
```diff
@@ -45,6 +45,12 @@ module Ameba
         subject.expand(["**/#{current_file_basename}", "**/#{current_file_basename}"])
           .should eq [current_file_path]
       end
+
+      it "raises an ArgumentError when the glob doesn't match any files" do
+        expect_raises(ArgumentError, "No files found matching foo/*") do
+          subject.expand(["foo/*"])
+        end
+      end
```
Comment (veelenga) [inline spec/ameba/glob_utils_spec.cr:53]: please make sure to add more test cases:

1. missing files in a glob (already added)
2. missing file
3. missing folder
4. empty folder
5. multiple arguments

File: src/ameba/glob_utils.cr (hunk 1 above)
Comment (stufro) [reply src/ameba/glob_utils.cr:27]: I originally had the assertion below `glob += "/**/*.cr" if File.directory?(glob)` however this causes the below error when running `crystal src/cli.cr`.

```
> crystal src/cli.cr                                                                                                                              
globs: ["**/*.cr"]
globs: ["lib"]
Error: No files found matching lib/**/*.cr
```

`expand` is called with `["lib"]` because it is one of the default rejected globs. But since it is empty the exception is raised.

Therefore I went down the route of only raising if the given directory doesn't exists, regardless of whether there are `.cr` files in the directory.

Interested in your thoughts on this.

File: spec/ameba/glob_utils_spec.cr (hunk 3 above)
Comment (stufro) [reply spec/ameba/glob_utils_spec.cr:53]: b2069ea - added extra test cases apart from for empty folder, as per above discussion.

File: src/ameba/glob_utils.cr (hunk 1 above)
Comment (veelenga) [reply src/ameba/glob_utils.cr:27]: @stufro I believe now the `!lib` glob has to be added to the default globs conditionally (i.e. excluded if exists). WDYT?

Comment (stufro) [reply src/ameba/glob_utils.cr:27]: Hm yes that's a good point. I guess I was assuming that every Crystal project will have lib/ but that's not a safe assumption. 

I'll have a look at conditionally adding !lib to the default globs.

Comment (stufro) [reply src/ameba/glob_utils.cr]: 7690074 - WDYT..? I'm not super happy with it but it's the best I could come up with

File: src/ameba/config.cr [hunk 4]
```diff
@@ -239,6 +239,13 @@ class Ameba::Config
     end
   end
 
+  private def default_globs
+    DEFAULT_GLOBS.select do |glob|
+      !Dir[glob].empty? ||
+        (glob.starts_with?("!") && !Dir["#{glob[1..-1]}/**/*.cr"].empty?)
+    end
+  end
```
Comment (veelenga) [inline src/ameba/config.cr:247]: I would probably remove `!lib` part from the DEFAULT_GLOBS and include it conditionally here (as an opposite to excluding it). Also, DEFAULT_GLOBS would probably have to be renamed:

```crystal
MATCHING_SOURCES_GLOB = "**/*.cr"
```

WDYT?

Comment (stufro) [reply src/ameba/config.cr]: Thank you! I've pushed eb60b25 using a similar approach to what you described.

Let me know if there's anything else we can improve me merging this :)

File: spec/ameba/config_spec.cr [hunk 5]
```diff
@@ -9,10 +9,20 @@ module Ameba
...
+          config.globs.should eq ["**/*.cr"]
+        end
+
+        it "only sets !lib as a default glob when there are .cr files in lib" do
+          File.touch "lib/shard.cr"
+          yml = YAML.parse "{}"
+          config = Config.new(yml)
+          config.globs.should eq ["**/*.cr", "!lib"]
+          File.delete "lib/shard.cr"
+        end
```
Comment (veelenga) [inline spec/ameba/config_spec.cr:25]: cleanup must be guaranteed in the tests:

```suggestion
        it "only sets !lib as a default glob when there are .cr files in lib" do
          File.touch "lib/shard.cr"
          yml = YAML.parse "{}"
          config = Config.new(yml)
          config.globs.should eq ["**/*.cr", "!lib"]
        ensure
          File.delete "lib/shard.cr"
        end
```

File: src/ameba/config.cr [hunk 6]
```diff
@@ -239,6 +236,12 @@ class Ameba::Config
     end
   end
 
+  private def default_globs
+    [SOURCES_GLOB].tap do |globs|
+      globs.push("!lib") if !Dir["lib/**/*.cr"].empty?
```
Comment (veelenga) [inline src/ameba/config.cr:241]:
```suggestion
      globs.push("!lib") unless Dir["lib/**/*.cr"].empty?
```

**Review by veelenga — COMMENTED** [review]
Just a few minor suggestions on my end. Rest LGTM

**Review by veelenga — APPROVED** [review]
LGTM

**Comment by Sija** [comment]
This release [broke](https://github.com/crystal-ameba/github-action/actions/runs/7324627583/job/19948387851) ameba's github action :( Also, I believe it's incorrect to raise for invalid paths in `Globs` or `Excluded` - these can be potentially empty or missing and neither of these should be a reason to raise an error and halt the execution of the whole program.

**Comment by stufro** [comment]
Ah sorry about this @Sija. I'm happy to have a look at resolving this but won't likely have time for several weeks so it might be best to revert this PR for now.

**Comment by Sija** [comment]
@stufro No worries, let's revert this PR for now and take care of it l8r.
